File: ios/Runner/QnScalePlugin.m
//
//  QnScalePlugin.m
//  Copy to ios/Runner/QnScalePlugin.m
//
//  Bridge between the Qingniu scale SDK and Flutter.
//
//  Channel names and event shapes match the Android bridge exactly,
//  so the Dart side does not care which platform it is running on.
//
//  Fails soft throughout: a scale that misbehaves sends an error
//  event, never an exception across the channel.
//

#import "QnScalePlugin.h"
#import <QNSDK/QNDeviceSDK.h>

@interface QnScalePlugin () <FlutterStreamHandler,
                             QNBleDeviceDiscoveryListener,
                             QNBleConnectionChangeListener,
                             QNScaleDataListener>

@property (nonatomic, strong) FlutterEventSink eventSink;
@property (nonatomic, strong) NSMutableDictionary<NSString *, QNBleDevice *> *devices;
@property (nonatomic, strong) QNBleDevice *currentDevice;
@property (nonatomic, strong) QNUser *user;

@end

@implementation QnScalePlugin

// Held so the instance survives registration — the channels keep only
// weak references to their handlers.
static QnScalePlugin *sharedPlugin = nil;

+ (void)registerWithMessenger:(NSObject<FlutterBinaryMessenger> *)messenger {
    QnScalePlugin *instance = [[QnScalePlugin alloc] init];
    sharedPlugin = instance;
    instance.devices = [NSMutableDictionary dictionary];

    FlutterMethodChannel *method =
        [FlutterMethodChannel methodChannelWithName:@"bmh/qn_scale"
                                    binaryMessenger:messenger];
    [method setMethodCallHandler:^(FlutterMethodCall *call, FlutterResult result) {
        [instance handleMethodCall:call result:result];
    }];

    FlutterEventChannel *events =
        [FlutterEventChannel eventChannelWithName:@"bmh/qn_scale_events"
                                  binaryMessenger:messenger];
    [events setStreamHandler:instance];
}

#pragma mark - Events

- (FlutterError *)onListenWithArguments:(id)arguments
                              eventSink:(FlutterEventSink)events {
    self.eventSink = events;
    return nil;
}

- (FlutterError *)onCancelWithArguments:(id)arguments {
    self.eventSink = nil;
    return nil;
}

- (void)send:(NSDictionary *)payload {
    // Sinks must be touched on the main thread; SDK callbacks make no
    // promises about which thread they arrive on.
    dispatch_async(dispatch_get_main_queue(), ^{
        if (self.eventSink) self.eventSink(payload);
    });
}

- (void)sendError:(NSString *)message {
    [self send:@{@"event": @"error", @"message": message ?: @"Unknown error"}];
}

#pragma mark - Methods

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
    NSDictionary *args = [call.arguments isKindOfClass:[NSDictionary class]]
        ? call.arguments : @{};

    if ([call.method isEqualToString:@"init"]) {
        [self initSdk:args result:result];
    } else if ([call.method isEqualToString:@"startScan"]) {
        [self startScan:result];
    } else if ([call.method isEqualToString:@"stopScan"]) {
        [self stopScan:result];
    } else if ([call.method isEqualToString:@"connect"]) {
        [self connect:args result:result];
    } else if ([call.method isEqualToString:@"disconnect"]) {
        [self disconnect:result];
    } else {
        result(FlutterMethodNotImplemented);
    }
}

- (void)initSdk:(NSDictionary *)args result:(FlutterResult)result {
    NSString *appId = args[@"appId"] ?: @"123456789";
    NSString *resource = args[@"configIos"] ?: @"123456789";

    NSString *file = [[NSBundle mainBundle] pathForResource:resource ofType:@"qn"];
    if (file == nil) {
        // Almost always means the .qn file is in the folder but not
        // ticked for the Runner target, so it never reached the bundle.
        [self sendError:[NSString stringWithFormat:
            @"Config file %@.qn is not in the app bundle.", resource]];
        result(@(NO));
        return;
    }

    [[QNBleApi sharedBleApi] initSdk:appId
                       firstDataFile:file
                            callback:^(NSError *error) {
        if (error) {
            [self sendError:[NSString stringWithFormat:
                @"SDK init failed: %@", error.localizedDescription]];
            result(@(NO));
        } else {
            result(@(YES));
        }
    }];
}

- (void)startScan:(FlutterResult)result {
    // Assigned as a property. The documentation describes
    // setBleDeviceDiscoveryListener:, which is the Android API — the
    // iOS SDK exposes these as weak properties instead.
    [QNBleApi sharedBleApi].discoveryListener = self;
    [[QNBleApi sharedBleApi] startBleDeviceDiscovery:^(NSError *error) {
        if (error) {
            [self sendError:[NSString stringWithFormat:
                @"Scan failed: %@", error.localizedDescription]];
        }
    }];
    result(@(YES));
}

- (void)stopScan:(FlutterResult)result {
    // "Discorvery" is spelled that way in the SDK header. Not a typo
    // here — correcting it does not compile.
    [[QNBleApi sharedBleApi] stopBleDeviceDiscorvery:^(NSError *error) {}];
    result(@(YES));
}

- (void)connect:(NSDictionary *)args result:(FlutterResult)result {
    NSString *mac = args[@"mac"];
    if (mac.length == 0) { result(@(NO)); return; }

    // iOS gives no MAC address for a peripheral, so a device cannot be
    // rebuilt from a string the way it can on Android. The object from
    // discovery is kept and reused instead.
    QNBleDevice *device = self.devices[mac];
    if (device == nil) {
        [self sendError:@"That scale is no longer in range. Scan again."];
        result(@(NO));
        return;
    }
    self.currentDevice = device;

    NSString *userId = args[@"userId"] ?: @"bmh_local_user";
    NSString *gender = args[@"gender"] ?: @"male";
    int height = [args[@"height"] intValue] ?: 170;
    NSString *birthdayStr = args[@"birthday"] ?: @"1990-01-01";

    NSDateFormatter *formatter = [[NSDateFormatter alloc] init];
    formatter.dateFormat = @"yyyy-MM-dd";
    NSDate *birthday = [formatter dateFromString:birthdayStr] ?: [NSDate date];

    self.user = [[QNBleApi sharedBleApi] buildUser:userId
                                            height:height
                                            gender:gender
                                          birthday:birthday
                                          callback:^(NSError *error) {
        if (error) {
            [self sendError:[NSString stringWithFormat:
                @"User: %@", error.localizedDescription]];
        }
    }];

    [QNBleApi sharedBleApi].connectionChangeListener = self;
    [QNBleApi sharedBleApi].dataListener = self;

    [[QNBleApi sharedBleApi] connectDevice:device
                                      user:self.user
                                  callback:^(NSError *error) {
        if (error) {
            [self sendError:[NSString stringWithFormat:
                @"Connect: %@", error.localizedDescription]];
        }
    }];
    result(@(YES));
}

- (void)disconnect:(FlutterResult)result {
    if (self.currentDevice) {
        [[QNBleApi sharedBleApi] disconnectDevice:self.currentDevice
                                         callback:^(NSError *error) {}];
    }
    self.currentDevice = nil;
    result(@(YES));
}

#pragma mark - Discovery

- (void)onDeviceDiscover:(QNBleDevice *)device {
    if (device == nil || device.mac == nil) return;
    self.devices[device.mac] = device;
    [self send:@{
        @"event": @"device",
        @"mac": device.mac ?: @"",
        @"name": device.name ?: @"Scale",
        @"rssi": device.RSSI ?: @(0),
        @"modelId": device.modeId ?: @"",
    }];
}

- (void)onStartScan {}
- (void)onStopScan {}

#pragma mark - Connection

- (void)onConnecting:(QNBleDevice *)device {
    [self send:@{@"event": @"connection", @"state": @"connecting"}];
}

- (void)onConnected:(QNBleDevice *)device {
    [self send:@{@"event": @"connection", @"state": @"connected"}];
}

- (void)onDisconnecting:(QNBleDevice *)device {}

- (void)onDisconnected:(QNBleDevice *)device {
    [self send:@{@"event": @"connection", @"state": @"disconnected"}];
}

- (void)onConnectError:(QNBleDevice *)device error:(NSError *)error {
    [self sendError:[NSString stringWithFormat:
        @"Connection error: %@", error.localizedDescription]];
}

- (void)onServiceSearchComplete:(QNBleDevice *)device {}

#pragma mark - Data

// Weight while the person is still settling. Streamed for the live
// display only, never stored — a moving number is not a measurement.
- (void)onGetUnsteadyWeight:(QNBleDevice *)device weight:(double)weight {
    [self send:@{@"event": @"weighing", @"weight": @(weight)}];
}

- (void)onGetScaleData:(QNBleDevice *)device data:(QNScaleData *)scaleData {
    NSMutableArray *items = [NSMutableArray array];
    for (QNScaleItemData *item in [scaleData getAllItem]) {
        [items addObject:@{
            @"type": @(item.type),
            @"name": item.name ?: @"",
            @"value": @(item.value),
            @"valueType": @(item.valueType),
        }];
    }

    double weight = 0;
    QNScaleItemData *weightItem = [scaleData getItem:QNScaleTypeWeight];
    if (weightItem) weight = weightItem.value;

    [self send:@{
        @"event": @"measurement",
        @"weight": @(weight),
        @"measureTime": @([scaleData.measureTime timeIntervalSince1970] * 1000),
        @"hmac": scaleData.hmac ?: @"",
        @"items": items,
    }];
}

- (void)onGetStoredScale:(QNBleDevice *)device data:(NSArray *)storedDataList {}

- (void)onGetElectric:(NSUInteger)electric device:(QNBleDevice *)device {}

- (void)onScaleStateChange:(QNBleDevice *)device
                scaleState:(QNScaleState)state {
    // Surfaced so the UI can say "measuring body composition" rather
    // than sitting on a spinner. Bioimpedance only reads with bare
    // feet, which is why the app says so before you step on.
    if (state == QNScaleStateBodyFat) {
        [self send:@{@"event": @"connection", @"state": @"measuring"}];
    }
}

- (void)onScaleEventChange:(QNBleDevice *)device
                scaleEvent:(QNScaleEvent)scaleEvent {}

@end

File: ios/Runner/QnScalePlugin.swift
import Flutter
import UIKit
import QNSDK

// ─────────────────────────────────────────────────────────
//  QINGNIU SCALE BRIDGE — iOS
//
//  Copy to: ios/Runner/QnScalePlugin.swift
//  Add to the Runner target in Xcode.
//
//  `import QNSDK` works because the Podfile uses use_frameworks! and
//  the pod declares static_framework, so CocoaPods builds it as a
//  module. Without use_frameworks! this would instead need a bridging
//  header line: #import "QNBleApi.h"
//
//  Mirrors the Kotlin bridge exactly: same channel names, same event
//  shapes, so the Dart side does not care which platform it is on.
//
//  Fails soft throughout. A scale that will not connect sends an error
//  event; it never crashes the host app.
// ─────────────────────────────────────────────────────────

class QnScalePlugin: NSObject, FlutterStreamHandler {

    static let methodChannel = "bmh/qn_scale"
    static let eventChannel = "bmh/qn_scale_events"

    private var sink: FlutterEventSink?
    private var currentDevice: QNBleDevice?
    private var user: QNUser?

    // ── REGISTRATION ────────────────────────────────────
    // Held so the instance is not deallocated the moment register
    // returns — the channels keep only weak references to handlers.
    private static var shared: QnScalePlugin?

    static func register(with messenger: FlutterBinaryMessenger) {
        let instance = QnScalePlugin()
        shared = instance

        let method = FlutterMethodChannel(
            name: methodChannel,
            binaryMessenger: messenger)
        method.setMethodCallHandler { call, result in
            instance.handle(call, result: result)
        }

        let events = FlutterEventChannel(
            name: eventChannel,
            binaryMessenger: messenger)
        events.setStreamHandler(instance)
    }

    // ── EVENTS ──────────────────────────────────────────
    func onListen(withArguments arguments: Any?,
                  eventSink events: @escaping FlutterEventSink)
        -> FlutterError? {
        sink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        return nil
    }

    private func send(_ payload: [String: Any?]) {
        DispatchQueue.main.async { [weak self] in
            self?.sink?(payload)
        }
    }

    private func sendError(_ message: String) {
        send(["event": "error", "message": message])
    }

    // ── METHODS ─────────────────────────────────────────
    private func handle(_ call: FlutterMethodCall,
                        result: @escaping FlutterResult) {
        switch call.method {
        case "init":       initSdk(call, result)
        case "startScan":  startScan(result)
        case "stopScan":   stopScan(result)
        case "connect":    connect(call, result)
        case "disconnect": disconnect(result)
        default:           result(FlutterMethodNotImplemented)
        }
    }

    private func initSdk(_ call: FlutterMethodCall,
                         _ result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any] ?? [:]
        let appId = args["appId"] as? String ?? "123456789"
        let resource = args["configIos"] as? String ?? "123456789"

        guard let path = Bundle.main.path(forResource: resource,
                                          ofType: "qn") else {
            // The usual cause is the .qn file not being ticked for the
            // Runner target in Xcode — it is in the folder but not in
            // the bundle.
            sendError("Config file \(resource).qn is not in the bundle.")
            result(false)
            return
        }

        QNBleApi.sharedBleApi().initSdk(appId, firstDataFile: path) {
            [weak self] error in
            if let error = error {
                self?.sendError("SDK init failed: \(error.localizedDescription)")
                result(false)
            } else {
                result(true)
            }
        }
    }

    private func startScan(_ result: @escaping FlutterResult) {
        QNBleApi.sharedBleApi().setBleDeviceDiscoveryListener(self)
        QNBleApi.sharedBleApi().startBleDeviceDiscovery { [weak self] error in
            if let error = error {
                self?.sendError("Scan failed: \(error.localizedDescription)")
            }
        }
        result(true)
    }

    private func stopScan(_ result: @escaping FlutterResult) {
        QNBleApi.sharedBleApi().stopBleDeviceDiscovery { _ in }
        result(true)
    }

    private func connect(_ call: FlutterMethodCall,
                         _ result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any] ?? [:]
        guard let mac = args["mac"] as? String, !mac.isEmpty else {
            result(false)
            return
        }

        let userId = args["userId"] as? String ?? "bmh_local_user"
        let gender = args["gender"] as? String ?? "male"
        let height = args["height"] as? Int ?? 170
        let birthdayStr = args["birthday"] as? String ?? "1990-01-01"

        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        let birthday = formatter.date(from: birthdayStr) ?? Date()

        user = QNBleApi.sharedBleApi().buildUser(
            userId,
            height: height,
            gender: gender,
            birthday: birthday,
            athleteType: 0) { [weak self] error in
                if let error = error {
                    self?.sendError("User: \(error.localizedDescription)")
                }
            }

        let device = currentDevice ?? QNBleApi.sharedBleApi().buildDevice(
            withMac: mac, name: "BioScale", modeId: "") { _ in }
        guard let device = device, let user = user else {
            sendError("Could not build a device for \(mac)")
            result(false)
            return
        }
        currentDevice = device

        QNBleApi.sharedBleApi().setBleConnectionChangeListener(self)
        QNBleApi.sharedBleApi().setDataListener(self)

        QNBleApi.sharedBleApi().connectDevice(device, user: user) {
            [weak self] error in
            if let error = error {
                self?.sendError("Connect: \(error.localizedDescription)")
            }
        }
        result(true)
    }

    private func disconnect(_ result: @escaping FlutterResult) {
        if let d = currentDevice {
            QNBleApi.sharedBleApi().disconnectDevice(d) { _ in }
        }
        currentDevice = nil
        result(true)
    }
}

// ── SDK CALLBACKS ───────────────────────────────────────
extension QnScalePlugin: QNBleDeviceDiscoveryListener {
    func onDeviceDiscover(_ device: QNBleDevice) {
        send([
            "event": "device",
            "mac": device.mac,
            "name": device.name,
            "rssi": device.rssi,
            "modelId": device.modeId,
        ])
    }
}

extension QnScalePlugin: QNBleConnectionChangeListener {
    func onConnecting(_ device: QNBleDevice) {
        send(["event": "connection", "state": "connecting"])
    }

    func onConnected(_ device: QNBleDevice) {
        send(["event": "connection", "state": "connected"])
    }

    func onDisconnecting(_ device: QNBleDevice) {}

    func onDisconnected(_ device: QNBleDevice) {
        send(["event": "connection", "state": "disconnected"])
    }

    func onConnectError(_ device: QNBleDevice, error: Error) {
        sendError("Connection error: \(error.localizedDescription)")
    }

    func onServiceSearchComplete(_ device: QNBleDevice) {}
}

extension QnScalePlugin: QNScaleDataListener {
    // Live weight while the person is still settling — for the
    // animation only, never stored.
    func onGetUnsteadyWeight(_ device: QNBleDevice, weight: Double) {
        send(["event": "weighing", "weight": weight])
    }

    func onGetScaleData(_ device: QNBleDevice, data: QNScaleData) {
        var items: [[String: Any?]] = []
        for item in data.getAllItem() {
            items.append([
                "type": item.type,
                "name": item.name,
                "value": item.value,
                "valueType": item.valueType,
            ])
        }
        send([
            "event": "measurement",
            "weight": data.getItemValue(1),
            "measureTime": data.measureTime.timeIntervalSince1970 * 1000,
            "hmac": data.hmac,
            "items": items,
        ])
    }

    func onGetStoredScale(_ device: QNBleDevice,
                          storedDataList: [QNScaleStoreData]) {}

    func onGetElectric(_ electric: UInt, device: QNBleDevice) {}

    func onScaleStateChange(_ device: QNBleDevice, scaleState: QNScaleState) {}

    func onScaleEventChange(_ device: QNBleDevice, scaleEvent: QNScaleEvent) {}
}
